File: packages/langgraph-bridge/jaxx_bridge/schemas.py
"""Schemas mirroring @jaxx/core TypeScript schemas (see packages/core/src/schemas.ts)."""

from __future__ import annotations

import re
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator

EventLevel = Literal["INFO", "WARN", "ERROR", "DONE", "AGENT", "GIT"]
EVENT_LEVELS = ("INFO", "WARN", "ERROR", "DONE", "AGENT", "GIT")


class Event(BaseModel):
    """Mirror of the canonical AGENT_LOG.jsonl entry."""

    ts: str
    lvl: EventLevel
    agent: str = Field(min_length=1)
    msg: str = Field(min_length=1)

    @field_validator("ts")
    @classmethod
    def ts_must_be_iso(cls, v: str) -> str:
        datetime.fromisoformat(v.replace("Z", "+00:00"))
        return v

    @classmethod
    def now(cls, lvl: EventLevel, agent: str, msg: str) -> "Event":
        return cls(ts=datetime.now().astimezone().isoformat(), lvl=lvl, agent=agent, msg=msg)


class RunRequest(BaseModel):
    goal: str = Field(min_length=1, max_length=2000)
    agent: str = Field(default="orchestrator", min_length=1, max_length=64)


class RunResponse(BaseModel):
    ok: bool
    events_appended: int
    summary: str


_NAME_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")


class SkillFrontmatter(BaseModel):
    """Mirror of the skill frontmatter schema (declarative metadata only)."""

    name: str = Field(min_length=1, max_length=64)
    description: str = Field(min_length=1, max_length=500)
    trigger: str = Field(min_length=1, max_length=200)
    allowedTools: list[str] = Field(default_factory=list)
    version: str = Field(default="0.1.0")

    @field_validator("version")
    @classmethod
    def semver(cls, v: str) -> str:
        if not re.match(r"^\d+\.\d+\.\d+$", v):
            raise ValueError("version must be semver")
        return v


def is_safe_name(name: str) -> bool:
    return bool(_NAME_RE.match(name))

File: packages/core/src/schemas.ts
import { z } from "zod";

/** Severity levels for control-plane events. */
export const EVENT_LEVELS = ["INFO", "WARN", "ERROR", "DONE", "AGENT", "GIT"] as const;
export type EventLevel = (typeof EVENT_LEVELS)[number];

/**
 * Canonical event schema. Every entry in AGENT_LOG.jsonl must satisfy this.
 * Mirrored by @jaxx/langgraph-bridge in Python.
 */
export const EventSchema = z.object({
  ts: z.string().datetime({ offset: true }),
  lvl: z.enum(EVENT_LEVELS),
  agent: z.string().min(1),
  msg: z.string().min(1),
});
export type Event = z.infer<typeof EventSchema>;

export const RepoConfigSchema = z.object({
  name: z.string().min(1),
  path: z.string().min(1),
  defaultBranch: z.string().min(1).default("main"),
});
export type RepoConfig = z.infer<typeof RepoConfigSchema>;

export const QualityConfigSchema = z
  .object({
    enabled: z.boolean().default(true),
    maxComplexity: z.number().int().positive().default(10),
    maxDuplicationRatio: z.number().min(0).max(1).default(0.05),
    exclude: z.array(z.string()).default([]),
  })
  .default({});
export type QualityConfig = z.infer<typeof QualityConfigSchema>;

/** frame.config.ts schema — the single whitelabel configuration surface. */
export const FrameConfigSchema = z.object({
  project: z.object({
    name: z.string().min(1),
    logoPath: z.string().optional(),
  }),
  theme: z
    .object({
      primaryColor: z.string().regex(/^#([0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/).default("#2563eb"),
      /** Optional page background override (defaults to the built-in dark surface). */
      backgroundColor: z.string().regex(/^#([0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/).optional(),
      borderRadius: z.string().default("0.5rem"),
    })
    .default({}),
  repos: z.array(RepoConfigSchema).default([]),
  docker: z.object({ containers: z.array(z.string()).default([]) }).default({}),
  ports: z.object({ dashboard: z.number().int().min(1).max(65535).default(3099) }).default({}),
  bridge: z
    .object({
      port: z.number().int().min(1).max(65535).default(3100),
    })
    .default({}),
  quality: QualityConfigSchema,
  tokenCountdown: z
    .object({
      enabled: z.boolean().default(false),
      /** Minutes until the countdown resets. */
      resetMinutes: z.number().int().positive().default(120),
      label: z.string().default("Token window"),
    })
    .default({}),
});
export type FrameConfig = z.infer<typeof FrameConfigSchema>;

/**
 * Skill frontmatter schema. Skills are Markdown files with YAML frontmatter.
 * NOTE: skill content is UNTRUSTED INPUT — see docs/security.md. Nothing here
 * is ever executed; `allowedTools` is declarative metadata only.
 */
export const SkillFrontmatterSchema = z.object({
  name: z.string().min(1).max(64).regex(/^[a-z0-9][a-z0-9._-]*$/i),
  description: z.string().min(1).max(500),
  trigger: z.string().min(1).max(200),
  allowedTools: z.union([z.array(z.string()), z.string()]).transform((v) =>
    typeof v === "string" ? v.split(",").map((s) => s.trim()).filter(Boolean) : v,
  ),
  version: z.string().regex(/^\d+\.\d+\.\d+$/).default("0.1.0"),
});
export type SkillFrontmatter = z.infer<typeof SkillFrontmatterSchema> & { allowedTools: string[] };

function pathIsTraversal(p: string): boolean {
  return p.split(/[\\/]/).includes("..");
}

export function parseEvent(raw: unknown): Event | null {
  const r = EventSchema.safeParse(raw);
  return r.success ? r.data : null;
}
